cpp/src/arrow/array/builder_nested.cc

```diff
@@ -99,6 +99,11 @@ Status ListBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
     RETURN_NOT_OK(value_builder_->FinishInternal(&items));
   }
 
+  // If the type hasn't been specified in the constructor, try to infer it
```

Comment:
Yes, this is for the case in which types are inferred at runtime.

Especially for nested builders and union builders, that's a very useful feature (potentially also for parsing CSV files and other operations where we don't know the types beforehand; sometimes it can be precomputed but espeically if there is a deeply nested data structure, that might be unnatural like for this PR). The way I have implemented it for this PR is by specifying a "nullptr" instead of a type, in which case the type will be computed upon Finish (and it will have the correct type including all the nested types). This is what this line is doing.

The way it is currently done in the PR shouldn't be merged, but we should have a good way to do this. There are two possibilities that come to my mind (they are all not super great, I hope you can come up with a better one):

- Introduce and special type (Type::DYNAMIC or TYPE::BUILDING) which only be a type of a Builder, not an Array. If the Builder has this type, the real type will be inferred upon Finish(). This is compatible with the current API, but introduces this awkward type (and we have to be careful people don't use it for normal Arrays).

- Change the builder API and remove the type() method. This might be too intrusive.

- Do it similar to the way I'm doing it right now, and have a check in the type() method that the type is not nullptr.

Let me know about your thoughts @pitrou @wesm @xhochy

Reply:
Another possibility I had considered is to have a boolean flag for the constructor of the builder that makes the type inference dynamic (in which case the type() method will check fail if it is called before the builder is finished).

Reply:
I realize you may have misunderstood my question above. To rephrase it: the type is _already_ inferred in the `ListBuilder` constructor, why is it needed to infer it again here?

Reply:
The point of this is to not infer the type in the constructors, but in the Finish method. For more complex array builders like UnionBuilder, that's the first time we can reasonably know what the type is.